### ISSUE TITLE
Memory Leak disposing oldProgram

### ISSUE THREAD

**Issue by cplepage** [microsoft/TypeScript#58137]
### 🔎 Search Terms

memory leak

### 🕗 Version & Regression Information

- This is a crash

### 💻 Code

```ts
// Your code here
```

### 🙁 Actual behavior

In safari (Webkit/JavaScriptCore). Old Program never gets GCed because of reference on it.

Memory build up really quickly.

### 🙂 Expected behavior

Unreferenced and gc

**Comment by cplepage** [comment]
Safari web inspector 

without fix:
<img width="935" alt="Screenshot 2024-04-09 at 3 25 25 PM" src="https://github.com/microsoft/TypeScript/assets/32472542/ff393e91-9e21-4f84-9e66-ebe3d96698e4">

with fix:
<img width="888" alt="Screenshot 2024-04-09 at 3 27 02 PM" src="https://github.com/microsoft/TypeScript/assets/32472542/93a203a4-8471-44b0-9fd8-1dc60a849dbf">

**Comment by RyanCavanaugh** [comment]
Accepting PRs to fix memory leaks, but we obviously can't really investigate something without a repro

**Comment by cplepage** [comment]
@RyanCavanaugh definitely, I will build a little something to make this issue reproducible  and figure out what could go on. But I will need some time. Please allow me 1-2 week and I will come back here with a clear workable playground.

**Comment by cplepage** [comment]
The easiest way to reproduce will be to run it as I encountered the issue.

### Requirements
* Git
* Node `>=18`
* MacOS (issue is only in Safari)

### Setting up

1. Clone the FullStacked Editor repo
2. Checkout the `typescript` branch
3. Pull all submodules
4. Build TypeScript
5. Install FullStacked Editor deps
6. Start with the MUI sample repo
7. Wait until the MUI sample project launches

Do all of this by copy/pasting the below snippet in Terminal. 
Some of these steps will take quite a while. 
Full process is up to 10 minutes.

```
git clone https://github.com/fullstackedorg/editor.git; cd editor
git checkout typescript
git submodule update --init --recursive --progress
cd lib/typescript; npm i; npm run build; cd ../..
npm i
npm start -- "https://github.com/fullstackedorg/editor-sample-mui.git?branch=ts"
```
*If at the end of the process you are not in Safari, close all popped up tabs and open http://localhost:9000 in Safari.

### Testing the issue

In Safari, in FullStacked Editor, in the `fullstackedorg/editor-sample-mui` project,

1. Open any `.ts`/`.tsx` file and wait for the TypeScript icon (top-right) to stop flashing.
2. Open up the debug console, go in the Timelines tab and start recording.
3. Edit the opened `.ts`/`.tsx` file constantly (hit enter at an interval of 1-2s) and monitor the memory. It should stay around 500-700mb

Now with your favorite code editor, open and comment out `lib/TypeScript/src/services/services.ts#1802`. Then, in Terminal, stop and rerun `npm start` and redo the three steps above. Check the memory climb nonstop (will go beyond 1GB).

**Comment by tmm1** [comment]
Friendly ping on repro

**Comment by cplepage** [comment]
Yes sorry about going radio silent here. 

Since discovering this issue and digging deep enough to believe the problem lays deep inside WebKit/JavaScriptCore, I simply patched it with my dumb fix https://github.com/fullstackedorg/editor/blob/main/build.ts#L9

To reproduce the issue, it needs a full in-browser setup of loading tons of files (node_modules) and requesting the language host repeatedly solely in-browser. I haven't found any browser IDE or code-editor which does that and is more than a playground which implies only a few files. 

If you have any example of fully in-browser code-editor/IDE that allows to load packages from say npm of added by .tgz, I would be glad to check if the issue is reproducible in any of those.

The project I'm working on, FullStacked, relies on that. For the last couple months, I've been working heavily on reaching a production ready stability, so I haven't put much thoughts on this issue. But you could try spinning up the project and simply compare in Safari the behaviour if you comment or not the code block previously mentioned here. 

There is no up-to-date contribution or how to build from source guide, so if you dive into that path, you'll probably need to figure out and modify some stuff. Although, I'm always available for any questions on the repo!

I will hopefully invest some time in documentation and said guides shortly. As soon as I can send a straightforward way to reproduce, I will do that here for sure. Also, send me links of app/IDE/code-editor that runs in-browser with the possibility of loading complete libraries with dependencies. I'll help as much as possible on the issue.

**Comment by tmm1** [comment]
Very interesting, thanks for the details.

There is another way to run tsserver with JSC, and in-fact I have observed similar memory growth issues in that environment- much more so than the equivalent code running in V8.

In VSCode:

![Image](https://github.com/user-attachments/assets/17641de6-6b95-4982-9732-118ce16c8009)

**Comment by cplepage** [comment]
Yes!! @tmm1 You are one clever guy!

The issue seems to be the same. The oldProgram being retained only in JSC.

Here the repo to test : https://github.com/cplepage/ts-leak.git

It simply `getSemanticDiagnostics` every 100ms on a basic MUI tsx file. The test runs for 60s while logging the memory usage. 

To run:
```
bun install
```

```
bun index.js
```

run with fix:
```
bun index.js --fix
```

run with node
```
node index.js
```
```
node index.js --fix
```

Here are the results on my end

| | no fix  | with fix  |
|---|---|---|
| bun | Memory after 1m: 1.05 GB<br />[max: 1.05 GB, min: 130 MB]  |  Memory after 1m: 126 MB<br />[max: 130 MB, min: 124 MB]  |
| node |  Memory after 1m: 383 MB<br />[max: 528 MB, min: 161 MB]  | Memory after 1m: 311 MB<br />[max: 528 MB, min: 167 MB] |

**Comment by tmm1** [comment]
Wonderful we have a repro! What a dramatic difference on the bun side.

@RyanCavanaugh wdyt

**Comment by tmm1** [comment]
My results:

||no fix|fix|
|-|-|-|
|bun 1.1.45|Memory after 1m: 1.7 GB [max: 1.7 GB, min: 148 MB]|Memory after 1m: 127 MB [max: 131 MB, min: 127 MB]|
|node 18.17.1|Memory after 1m: 445 MB [max: 536 MB, min: 194 MB]|Memory after 1m: 304 MB [max: 534 MB, min: 190 MB]|
|node 20.18.1|Memory after 1m: 518 MB [max: 518 MB, min: 193 MB]|Memory after 1m: 511 MB [max: 524 MB, min: 193 MB]|

**Comment by maschwenk** [comment]
That repro is very interesting. We have a similar problem where specifically on an app that _also_ depends on 

```
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0",
```

Can consume 9-10 GB in < a minute. I don't know if it'd technically be called a leak because it does not grow indefinitely though. 

I've tried applying this fix and does not seem to help, so maybe unrelated, but I'm sort of curious if the choice of mui/@emotion was on purpose? I've also noticed that even after all files are closed in VScode, memory will be pinned to that 9-10GB number indefinitely (which *does* seem a bit leaky?). This https://github.com/cplepage/ts-leak repro has inspired me to maybe make my own.

**Comment by jakebailey** [comment]
> I've also noticed that even after all files are closed in VScode, memory will be pinned to that 9-10GB number indefinitely (which does seem a bit leaky?).

It's doing that intentionally (the keeping in memory part, not the 10GB part), since you may reopen a file. If you close all your files, then open a file, it's not so good to have that be a complete restart.

That being said, I don't know how you can possibly see 9-10GB of memory usage; we have a hard cap when run via VS Code of 4GB, but maybe you have instructed VS Code to use the "real" node instead, but then surely that's a different bug to this JSC problem.

> This [https://github.com/cplepage/ts-leak](https://github.com/cplepage/ts-leak?rgh-link-date=2025-01-23T15%3A50%3A03.000Z) repro has inspired me to maybe make my own.

More repros are of course, good.

**Comment by jakebailey** [comment]
FWIW I still do not think #58138 itself is the right fix, just from an implementation perspective; it's possible there's a similar change that would work fine. But I really do wonder if this is something JSC should be fixing given V8 does not have this problem.

Note that in https://github.com/microsoft/TypeScript/issues/58137#issuecomment-2605899735, Node's (V8's) min/max memory usage is the same before and after the change (I would not focus on a specific measurement since that really depends on when the last GC run ran among other things).

**Comment by jakebailey** [comment]
I'm looking at the bun thing, but it sounds like your issue is likely unrelated unless you're applying #58138 and finding a benefit.

mui / emotion are famously some of the slowest types out there, so it doesn't surprise me too much; in your case, I would try and verify that your packages use the same tsconfig options to maximize reuse, and also consider using some of the "make tsserver load less with downsides" options:

- https://www.typescriptlang.org/tsconfig/#disableReferencedProjectLoad
- https://www.typescriptlang.org/tsconfig/#disableSolutionSearching
- https://www.typescriptlang.org/tsconfig/#disableSourceOfProjectReferenceRedirect

**Comment by jakebailey** [comment]
I really don't like the fix, but I cannot find anything that works. #61034 is my version of #51838 and some additional modifications which make me feel less scared to accept the hack. Let me know if it still works. It should, and the repro at least let me test it (thank you for making one).

**Comment by jakebailey** [comment]
Actually, I found the actual problem, #61034 now contains a fix for that.

**Comment by tmm1** [comment]
> Note that in [#58137 (comment)](https://github.com/microsoft/TypeScript/issues/58137#issuecomment-2605899735), Node's (V8's) min/max memory usage is the same before and after the change (I would not focus on a specific measurement since that really depends on when the last GC run ran among other things).

Good point, if you run with `node --expose-gc` and add explicit `global.gc()` calls before the measurements, the usage is completely flat (just like bunjs is after the patch).

**Comment by jakebailey** [comment]
Yep, that's exactly what I did locally when testing.

**Comment by tmm1** [comment]
> repro has inspired me to maybe make my own.

Would be great if we can find a repro and make a new issue.

I agree the current behavior wrt to memory usage on large typescript projects is far from ideal.


> we have a hard cap when run via VS Code of 4GB

Is this referring to the default max-old-space limit, or the pointer compression limit?


> It's doing that intentionally (the keeping in memory part, not the 10GB part), since you may reopen a file. If you close all your files, then open a file, it's not so good to have that be a complete restart.

Has there been any consideration to using some kind of LRU, instead of an unbounded map here:

https://github.com/microsoft/TypeScript/blob/0e371c4bf587e1254e5810d88ae055073b18ab0b/src/server/editorServices.ts#L1291
```
    readonly filenameToScriptInfo: Map<Path, ScriptInfo> = new Map();
```

I can observe this growing as I navigate into files and dependencies, and sticking around longer after I care about those files. Eventually memory will hit the defined limits, which results in a very poor experience as GC and cpu usage increase until an eventual OOM.

Here for instance 500mb is retained by this map, and there is 1.6G worth of SourceFileObject

<img width="1521" alt="Image" src="https://github.com/user-attachments/assets/b75b70d5-4806-44a4-9398-f161692fabf4" />

**Comment by jakebailey** [comment]
> Is this referring to the default max-old-space limit, or the pointer compression limit?

Pointer compression limit.

> Has there been any consideration to using some kind of LRU, instead of an unbounded map here:

I would suggest making a new issue; that seems unrelated to the leak from this thread. In your screenshot, there are only 4000 entries or so; that is the number of files in the VS Code project; it's not that large.
